File: go/Zadanie6/experiment.go
// experiment.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type ExperimentResult struct {
	MinLength  int
	MaxLength  int
	AvgLength  float64
	GenTime    int64
	InsertTime int64
	TableSize  int
	LoadFactor float64
	Method     CollisionResolution
}

type HashTableExperiment struct{}

func (hte *HashTableExperiment) RunSingleExperiment(N int, method CollisionResolution) ExperimentResult {
	result := ExperimentResult{}
	result.Method = method

	if method == CHAINING {
		result.TableSize = max(1, N/100)
	} else {
		result.TableSize = max(1, N*2)
	}

	result.LoadFactor = float64(N) / float64(result.TableSize)

	timer := &Timer{}
	rg := &RandomGenerator{}

	timer.Start()
	randomNumbers := rg.GenerateRandomNumbers(N, 0, 1000000)
	result.GenTime = timer.ElapsedMillis()

	hashTable := NewHashTable(result.TableSize, method)

	timer.Start()
	for i := 0; i < randomNumbers.Length(); i++ {
		num, _ := strconv.Atoi(randomNumbers.Get(i))
		hashTable.Insert(num)
	}
	result.InsertTime = timer.ElapsedMillis()

	minLen, maxLen, avgLen := hashTable.GetStatistics()
	result.MinLength = minLen
	result.MaxLength = maxLen
	result.AvgLength = avgLen

	randomNumbers.Destroy()

	return result
}

func (hte *HashTableExperiment) RunComparisonExperiment(N int) {
	fmt.Printf("=== Сравнительный эксперимент для N = %d ===\n", N)

	rg := &RandomGenerator{}
	randomNumbers := rg.GenerateRandomNumbers(N, 0, 1000000)

	tableSizes := []int{N / 200, N / 100, N / 50, N / 20}

	for _, tableSize := range tableSizes {
		if tableSize < 1 {
			tableSize = 1
		}

		hashTable := NewHashTable(tableSize, CHAINING)

		for i := 0; i < randomNumbers.Length(); i++ {
			num, _ := strconv.Atoi(randomNumbers.Get(i))
			hashTable.Insert(num)
		}

		minLen, maxLen, avgLen := hashTable.GetStatistics()

		fmt.Printf("Размер таблицы: %6d | Заполнение: %8.2f | Цепочки: мин=%3d, макс=%3d, средняя=%6.2f\n",
			tableSize, float64(N)/float64(tableSize), minLen, maxLen, avgLen)
	}

	randomNumbers.Destroy()
	fmt.Printf("\n%s\n\n", strings.Repeat("=", 50))
}

func (hte *HashTableExperiment) RunMethodComparisonExperiment(N int) {
	fmt.Printf("=== Сравнение методов разрешения коллизий для N = %d ===\n", N)

	methods := []CollisionResolution{CHAINING, LINEAR_PROBING, QUADRATIC_PROBING}
	methodNames := []string{"Метод цепочек", "Линейное пробирование", "Квадратичное пробирование"}

	for i := 0; i < 3; i++ {
		result := hte.RunSingleExperiment(N, methods[i])

		fmt.Printf("%s:\n", methodNames[i])
		fmt.Printf("  Время вставки: %d мс\n", result.InsertTime)
		fmt.Printf("  Цепочки: мин=%d, макс=%d, средняя=%.2f\n",
			result.MinLength, result.MaxLength, result.AvgLength)
		fmt.Printf("  Коэф. заполнения: %.2f\n\n", result.LoadFactor)
	}

	fmt.Printf("%s\n\n", strings.Repeat("=", 50))
}

func (hte *HashTableExperiment) PrintResult(result ExperimentResult, N int) {
	methodName := ""
	switch result.Method {
	case CHAINING:
		methodName = "Метод цепочек"
	case LINEAR_PROBING:
		methodName = "Линейное пробирование"
	case QUADRATIC_PROBING:
		methodName = "Квадратичное пробирование"
	}

	fmt.Printf("=== Эксперимент для N = %d (%s) ===\n", N, methodName)
	fmt.Printf("Размер хеш-таблицы: %d\n", result.TableSize)
	fmt.Printf("Коэффициент заполнения: %.2f\n", result.LoadFactor)
	fmt.Printf("Время генерации чисел: %d мс\n", result.GenTime)
	fmt.Printf("Время вставки: %d мс\n", result.InsertTime)
	fmt.Printf("\nРезультаты:\n")
	fmt.Printf("Самая короткая цепочка: %d\n", result.MinLength)
	fmt.Printf("Самая длинная цепочка: %d\n", result.MaxLength)
	fmt.Printf("Средняя длина непустой цепочки: %.2f\n", result.AvgLength)
}

func (hte *HashTableExperiment) PrintDetailedStatistics(hashTable *HashTable) {
	hashTable.PrintDetailedStatistics()
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: с++/Zadanie6/Zadanie6/experiment.cpp
#include "experiment.h"
#include <iostream>
#include <iomanip>

// Проведение одиночного эксперимента для N элементов
ExperimentResult HashTableExperiment::runSingleExperiment(int N, CollisionResolution method) {
    ExperimentResult result;
    result.method = method;

    // Оптимальные коэффициенты заполнения для разных методов
    if (method == CollisionResolution::CHAINING) {
        result.tableSize = std::max(1, N / 100);  // a = 1.0 - нормально для цепочек
    }
    else {
        // Для открытой адресации используем меньший коэффициент заполнения
        result.tableSize = std::max(1, N * 2);    // a = 0.5 - оптимально для открытой адресации
    }

    result.loadFactor = static_cast<double>(N) / result.tableSize;

    // Генерация случайных чисел и замер времени
    Timer timer;
    timer.start();
    Array* randomNumbers = RandomGenerator::generateRandomNumbers(N);
    result.genTime = timer.elapsedMillis();

    // Создание хеш-таблицы с выбранным методом
    HashTable hashTable(result.tableSize, method);

    // Заполнение таблицы и замер времени вставки
    timer.start();
    for (int i = 0; i < length(*randomNumbers); ++i) {
        int num = std::stoi(get(*randomNumbers, i));
        hashTable.insert(num);
    }
    result.insertTime = timer.elapsedMillis();

    // Сбор статистики по цепям
    hashTable.getStatistics(result.minLength, result.maxLength, result.avgLength);

    // Очистка памяти
    destroy(*randomNumbers);
    delete randomNumbers;

    return result;
}

// Проведение сравнительного эксперимента с разными размерами таблицы
void HashTableExperiment::runComparisonExperiment(int N) {
    std::cout << "=== Сравнительный эксперимент для N = " << N << " ===\n";

    // Генерация общих данных для всех тестов
    Array* randomNumbers = RandomGenerator::generateRandomNumbers(N);

    // Различные размеры таблицы для сравнения
    int tableSizes[] = { N / 200, N / 100, N / 50, N / 20 };

    // Тестирование каждого размера таблицы
    for (int tableSize : tableSizes) {
        if (tableSize < 1) tableSize = 1;

        HashTable hashTable(tableSize, CollisionResolution::CHAINING);

        // Заполнение таблицы
        for (int i = 0; i < length(*randomNumbers); ++i) {
            int num = std::stoi(get(*randomNumbers, i));
            hashTable.insert(num);
        }

        int minLength, maxLength;
        double avgLength;
        hashTable.getStatistics(minLength, maxLength, avgLength);

        // Вывод результатов сравнения
        std::cout << "Размер таблицы: " << std::setw(6) << tableSize
            << " | Заполнение: " << std::setw(8) << std::fixed << std::setprecision(2)
            << static_cast<double>(N) / tableSize
            << " | Цепочки: мин=" << std::setw(3) << minLength
            << ", макс=" << std::setw(3) << maxLength
            << ", средняя=" << std::setw(6) << std::setprecision(2) << avgLength << "\n";
    }

    // Очистка
    destroy(*randomNumbers);
    delete randomNumbers;
    std::cout << "\n" << std::string(50, '=') << "\n\n";
}

// Сравнение разных методов разрешения коллизий
void HashTableExperiment::runMethodComparisonExperiment(int N) {
    std::cout << "=== Сравнение методов разрешения коллизий для N = " << N << " ===\n";

    CollisionResolution methods[] = {
        CollisionResolution::CHAINING,
        CollisionResolution::LINEAR_PROBING,
        CollisionResolution::QUADRATIC_PROBING
    };

    const char* methodNames[] = {
        "Метод цепочек",
        "Линейное пробирование",
        "Квадратичное пробирование"
    };

    for (int i = 0; i < 3; ++i) {
        ExperimentResult result = runSingleExperiment(N, methods[i]);

        std::cout << methodNames[i] << ":\n";
        std::cout << "  Время вставки: " << result.insertTime << " мс\n";
        std::cout << "  Цепочки: мин=" << result.minLength
            << ", макс=" << result.maxLength
            << ", средняя=" << std::fixed << std::setprecision(2) << result.avgLength << "\n";
        std::cout << "  Коэф. заполнения: " << result.loadFactor << "\n\n";
    }

    std::cout << std::string(50, '=') << "\n\n";
}

// Вывод результатов одиночного эксперимента
void HashTableExperiment::printResult(const ExperimentResult& result, int N) {
    const char* methodName = "";
    switch (result.method) {
    case CollisionResolution::CHAINING:
        methodName = "Метод цепочек";
        break;
    case CollisionResolution::LINEAR_PROBING:
        methodName = "Линейное пробирование";
        break;
    case CollisionResolution::QUADRATIC_PROBING:
        methodName = "Квадратичное пробирование";
        break;
    }

    std::cout << "=== Эксперимент для N = " << N << " (" << methodName << ") ===\n";
    std::cout << "Размер хеш-таблицы: " << result.tableSize << "\n";
    std::cout << "Коэффициент заполнения: " << std::fixed << std::setprecision(2)
        << result.loadFactor << "\n";
    std::cout << "Время генерации чисел: " << result.genTime << " мс\n";
    std::cout << "Время вставки: " << result.insertTime << " мс\n";
    std::cout << "\nРезультаты:\n";
    std::cout << "Самая короткая цепочка: " << result.minLength << "\n";
    std::cout << "Самая длинная цепочка: " << result.maxLength << "\n";
    std::cout << "Средняя длина непустой цепочки: " << std::fixed << std::setprecision(2)
        << result.avgLength << "\n";
}

// Вывод детальной статистики хеш-таблицы
void HashTableExperiment::printDetailedStatistics(const HashTable& hashTable) {
    hashTable.printDetailedStatistics();
}